runtime/src/iree/builtins/ukernel/tools/mmt4d_test.c

```diff
@@ -167,6 +167,27 @@ static void iree_mmt4d_reference_innerloop_s8s8s32(
   *out_ptr = acc;
 }
 
+static void iree_mmt4d_reference_innerloop_s8s4s32(
+    int32_t* out_ptr, const int8_t* lhs_ptr, const int8_t* rhs_ptr,
+    const iree_uk_mmt4d_params_t* params) {
+  // K0 must be even.
+  IREE_UK_ASSERT(!(params->K0 % 2));
+  iree_uk_int16_t K0half = params->K0 / 2;
+  int32_t acc = params->flags & IREE_UK_FLAG_MMT4D_ACCUMULATE ? *out_ptr : 0;
+  for (iree_uk_index_t k = 0; k < params->K; ++k) {
+    // As K0 must be even, we 2x-unroll the K0 loop, writing a 2D dot product.
+    for (iree_uk_index_t k0h = 0; k0h < K0half; ++k0h) {
+      int32_t lhs_0 = lhs_ptr[k * params->M0 * params->K0 + 2 * k0h];
+      int32_t lhs_1 = lhs_ptr[k * params->M0 * params->K0 + 2 * k0h + 1];
+      int8_t rhs_byte = rhs_ptr[k * params->N0 * K0half + k0h];
+      int32_t rhs_0 = rhs_byte & 0xf;
```

Comment:
I was curious since we discussed the signed/unsigned discrepancy in the arm64 kernel code, if the `mmt4d_test` returned success, maybe that meant there was a corresponding bug in the reference code it's comparing against, and yes, this also is performing zero-extension not sign-extension, so this reference code is a `iree_mmt4d_reference_innerloop_s8u4s32` not a `iree_mmt4d_reference_innerloop_s8s4s32`.